{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# implementation of the naive exact matching algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def naive(p, t):\n",
    "    occurrences = []\n",
    "    for i in range(len(t) - len(p) + 1):  # loop over alignments\n",
    "        match = True\n",
    "        for j in range(len(p)):  # loop over characters\n",
    "            if t[i+j] != p[j]:  # compare characters\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurrences.append(i)  # all chars matched; record\n",
    "    return occurrences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### function that takes a DNA string and returns its reverse complement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def reverseComplement(s):\n",
    "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A', 'N': 'N'}\n",
    "    t = ''\n",
    "    for base in s:\n",
    "        t = complement[base] + t\n",
    "    return t"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### function that parses a DNA reference genome from a file in the FASTA format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### function that parses the read and quality strings from a FASTQ file containing sequencing reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def readFastq(filename):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    with open(filename) as fh:\n",
    "        while True:\n",
    "            fh.readline()  # skip name line\n",
    "            seq = fh.readline().rstrip()  # read base sequence\n",
    "            fh.readline()  # skip placeholder line\n",
    "            qual = fh.readline().rstrip() # base quality line\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "    return sequences, qualities"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### now one can implement naive exact matching algorithm that is strand-aware. (RC stads now for the reverse complement)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#from reversecomp import *\n",
    "#from naive import *\n",
    "\n",
    "def naive(p, t):\n",
    "    occurrences = []\n",
    "    for i in range(len(t) - len(p) + 1):  # loop over alignments\n",
    "        match = True\n",
    "        for j in range(len(p)):  # loop over characters\n",
    "            if t[i+j] != p[j]:  # compare characters\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurrences.append(i)  # all chars matched; record\n",
    "    return occurrences\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def naive_with_rc(p,t):\n",
    "    occur_for = naive(p,t)\n",
    "    occur_rev = naive(reverseComplement(p),t)\n",
    "    occurrences = []\n",
    "    for i in occur_for:\n",
    "        if i not in occurrences:\n",
    "            occurrences.append(i)\n",
    "    for i in occur_rev:\n",
    "        if i not in occurrences:\n",
    "            occurrences.append(i)\n",
    "\n",
    "    return(occurrences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Tests\n",
    "p = 'CCC'\n",
    "ten_as = 'AAAAAAAAAA'\n",
    "t = ten_as + 'CCC' + ten_as + 'GGG' + ten_as\n",
    "assert(naive_with_rc(p, t)) == [10,23]\n",
    "\n",
    "p = 'CGCG'\n",
    "t = ten_as + 'CGCG' + ten_as + 'CGCG' + ten_as\n",
    "assert(naive_with_rc(p, t)) == [10,24]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2020-04-06 19:54:14--  https://d28rh4a8wq0iu5.cloudfront.net/ads1/data/lambda_virus.fa\n",
      "Resolving d28rh4a8wq0iu5.cloudfront.net... 13.224.89.191, 13.224.89.79, 13.224.89.30, ...\n",
      "Connecting to d28rh4a8wq0iu5.cloudfront.net|13.224.89.191|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 49270 (48K) [application/octet-stream]\n",
      "Saving to: 'lambda_virus.fa.1'\n",
      "\n",
      "lambda_virus.fa.1   100%[=====================>]  48.12K  --.-KB/s   in 0.04s  \n",
      "\n",
      "2020-04-06 19:54:14 (1.17 MB/s) - 'lambda_virus.fa.1' saved [49270/49270]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# now doenwnload and parse lambda genome for exrecise\n",
    "\n",
    "!wget --no-check https://d28rh4a8wq0iu5.cloudfront.net/ads1/data/lambda_virus.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# \n",
    "lambda_genome = readGenome('lambda_virus.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# occurrences: 306\n"
     ]
    }
   ],
   "source": [
    "occurrences = naive_with_rc('AGGT', lambda_genome)\n",
    "#print(occurrences)\n",
    "print('# occurrences: %d' % len(occurrences))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# occurrences: 195\n"
     ]
    }
   ],
   "source": [
    "occurrences = naive_with_rc('TTAA', lambda_genome)\n",
    "#print(occurrences)\n",
    "print('# occurrences: %d' % len(occurrences))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "offset of leftmost occurrence: 26028\n"
     ]
    }
   ],
   "source": [
    "occurrences = naive_with_rc('ACTAAGT', lambda_genome)\n",
    "print('offset of leftmost occurrence: %d' % min(occurrences))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "offset of leftmost occurrence: 450\n"
     ]
    }
   ],
   "source": [
    "occurrences = naive_with_rc('AGTCGA', lambda_genome)\n",
    "print('offset of leftmost occurrence: %d' % min(occurrences))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Naive pattern matching with indexes of up to 2 mismatches\n",
    "\n",
    "def naive_2mm(p, t):\n",
    "    occurrences = []\n",
    "    for i in range(len(t) - len(p) + 1):  # loop over alignments\n",
    "        mismatches = 0\n",
    "        match = True\n",
    "        for j in range(len(p)):  # loop over characters\n",
    "            if t[i+j] != p[j]:  # compare characters\n",
    "                if mismatches < 2:\n",
    "                    mismatches += 1\n",
    "                else:\n",
    "                    match = False\n",
    "                    break\n",
    "        if match:\n",
    "            occurrences.append(i)  # all chars matched; record\n",
    "    return occurrences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# occurrences: 191\n"
     ]
    }
   ],
   "source": [
    "occurrences = naive_2mm('TTCAAGCC', lambda_genome)\n",
    "#print(occurrences)\n",
    "print('# occurrences: %d' % len(occurrences))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "offset of leftmost occurrence: 49\n"
     ]
    }
   ],
   "source": [
    "occurrences = naive_2mm('AGGAGGTT', lambda_genome)\n",
    "print('offset of leftmost occurrence: %d' % min(occurrences))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2020-04-06 20:27:32--  https://d28rh4a8wq0iu5.cloudfront.net/ads1/data/ERR037900_1.first1000.fastq\n",
      "Resolving d28rh4a8wq0iu5.cloudfront.net... 13.224.89.208, 13.224.89.191, 13.224.89.79, ...\n",
      "Connecting to d28rh4a8wq0iu5.cloudfront.net|13.224.89.208|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 241626 (236K) [application/octet-stream]\n",
      "Saving to: 'ERR037900_1.first1000.fastq'\n",
      "\n",
      "ERR037900_1.first10 100%[=====================>] 235.96K  1007KB/s   in 0.2s   \n",
      "\n",
      "2020-04-06 20:27:33 (1007 KB/s) - 'ERR037900_1.first1000.fastq' saved [241626/241626]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# now download human genome and test same functions \n",
    "\n",
    "\n",
    "!wget --no-check https://d28rh4a8wq0iu5.cloudfront.net/ads1/data/ERR037900_1.first1000.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def readFastq(filename):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    with open (filename) as fh:\n",
    "        while True:\n",
    "            fh.readline()\n",
    "            seq = fh.readline().rstrip()\n",
    "            fh.readline()\n",
    "            qual = fh.readline().rstrip()\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "    return sequences, qualities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "human_reads, qualities = readFastq('ERR037900_1.first1000.fastq')\n",
    "# print(human_reads, qualities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def phred33ToQ(qualities):\n",
    "    return ord(qualities) - 33"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 0, 17723, 0, 2, 11, 11, 28, 23, 55, 100, 111, 86, 174, 185, 272, 317, 259, 390, 1523, 2782, 762, 286, 413, 403, 538, 351, 694, 971, 777, 1024, 1449, 1341, 1312, 1916, 2233, 3025, 4043, 6640, 45696, 2074, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n"
     ]
    }
   ],
   "source": [
    "def createHist(qualities):\n",
    "    hist = [0] * 100\n",
    "    for qualities in qualities:\n",
    "        for phred in qualities:\n",
    "            q = phred33ToQ(phred)\n",
    "            hist[q] += 1\n",
    "    return hist\n",
    "h = createHist(qualities)\n",
    "print(h)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYoAAAEACAYAAACtVTGuAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAExtJREFUeJzt3X+snuVdx/H3h1ZgboCgtkdbGCxQhPljY6ZTF7OjLDC2\nBPhnTRd1zNV/BB1xxqzdP3T/uEFirEYhMeIoZLMyzEaXEehIPVmWMKm6CY4KTSZIyzgoFYyaLJR9\n/eO5u94967nO057n9Px43q+k4T7fc133ue77PDyf57ru+3lOqgpJkmZzxmIPQJK0tBkUkqQmg0KS\n1GRQSJKaDApJUpNBIUlqGiookjyb5J+TfCPJ413t/CR7kjyd5JEk5/Xab0tyIMn+JNf06lcleSLJ\nM0l29OpnJtnV9XksyUWjPEhJ0qkbdkbxPWCyqt5eVRu72lbg0aq6HNgLbANIciWwCbgCuA64M0m6\nPncBW6pqA7AhybVdfQtwuKouA3YAd8zzuCRJIzJsUOQEbW8AdnbbO4Ebu+3rgV1VdaSqngUOABuT\nTADnVNW+rt29vT79fT0AXH0yByFJWjjDBkUBX0myL8lvdbW1VTUNUFUvAmu6+jrg+V7fQ11tHXCw\nVz/Y1Y7rU1WvA68kueAkj0WStABWD9nuXVX1nSQ/DuxJ8jSD8Ogb5WeBZO4mkqTTYaigqKrvdP/9\njyRfBDYC00nWVtV0t6z0Utf8EHBhr/v6rjZbvd/nhSSrgHOr6vDMcSTxg6kk6RRU1Sm/AJ9z6SnJ\nDyd5U7f9RuAa4ElgN/DhrtlNwIPd9m5gc3cn0yXApcDj3fLUq0k2dhe3PzSjz03d9gcYXBw/oary\nXxW33Xbboo9hqfzzXHguPBftf/M1zIxiLfCF7tX8auCzVbUnyT8A9yf5CPAcgzudqKqnktwPPAW8\nBtxcx0Z6C3APcDbwUFU93NXvBu5LcgB4Gdg87yOTJI3EnEFRVf8GvO0E9cPAe2bp8yngUyeo/yPw\nMyeof5cuaCRJS4vvzF6mJicnF3sIS4bn4hjPxTGei9HJKNavTpcktZzGK0lLQRJqIS9mS5LGm0Eh\nSWoyKCRJTQaFJKnJoJAkNRkUkqQmg0KS1GRQSJKaDApJUpNBIUlqMigkSU0GhSSpyaCQJDUZFFqW\nJiYuJgkTExcv9lCkFc+PGdeyNPhrugVkJH/qUVrJ/JhxSdKCMigkSU0GhSSpyaCQJDUZFJKkJoNC\nktRkUEiSmgwKSVKTQSFJajIoJElNBoUkqcmgkCQ1GRSSpCaDQpLUZFBIkpoMCklSk0EhSWoyKCRJ\nTQaFJKlp6KBIckaSf0qyu/v6/CR7kjyd5JEk5/XabktyIMn+JNf06lcleSLJM0l29OpnJtnV9Xks\nyUWjOkBJ0vyczIziVuCp3tdbgUer6nJgL7ANIMmVwCbgCuA64M4kR/+o913AlqraAGxIcm1X3wIc\nrqrLgB3AHad4PJKkERsqKJKsB94H/GWvfAOws9veCdzYbV8P7KqqI1X1LHAA2JhkAjinqvZ17e7t\n9env6wHg6pM/FEnSQhh2RvHHwB8A1autrappgKp6EVjT1dcBz/faHepq64CDvfrBrnZcn6p6HXgl\nyQXDH4YkaaGsnqtBkvcD01X1zSSTjabV+N7Jymzf2L59+/e3JycnmZycHOGPlaTlb2pqiqmpqZHt\nL1Xt5/ckfwj8OnAEeANwDvAF4OeByaqa7paV/q6qrkiyFaiqur3r/zBwG/Dc0TZdfTPw7qr67aNt\nqurvk6wCvlNVa2YMhSQ113g1HgaXvQoIPiaktiRU1awvwOcy59JTVX2iqi6qqrcAm4G9VfUbwJeA\nD3fNbgIe7LZ3A5u7O5kuAS4FHu+Wp15NsrG7uP2hGX1u6rY/wODiuCRpCZhz6anh08D9ST7CYLaw\nCaCqnkpyP4M7pF4Dbu5NA24B7gHOBh6qqoe7+t3AfUkOAC8zCCRJ0hIw59LTUuLSk45y6Uka3oIv\nPUmSxptBIUlqMigkSU0GhSSpyaCQJDUZFJKkJoNCktRkUEiSmgwKSVKTQSFJajIoJElNBoUkqcmg\nkCQ1GRSSpCaDQpLUZFBIkpoMCklSk0EhSWoyKCRJTQaFJKnJoJAkNRkUkqQmg0KS1GRQSJKaDApJ\nUpNBIUlqMigkSU0GhSSpyaCQJDUZFJKkJoNCktRkUEiSmgwKSVKTQSFJajIoJElNcwZFkrOS/H2S\nbyR5MsltXf38JHuSPJ3kkSTn9fpsS3Igyf4k1/TqVyV5IskzSXb06mcm2dX1eSzJRaM+UEnSqZkz\nKKrqu8CvVNXbgbcB1yXZCGwFHq2qy4G9wDaAJFcCm4ArgOuAO5Ok291dwJaq2gBsSHJtV98CHK6q\ny4AdwB2jOkBJ0vwMtfRUVf/XbZ4FrAYKuAHY2dV3Ajd229cDu6rqSFU9CxwANiaZAM6pqn1du3t7\nffr7egC4+pSORpI0ckMFRZIzknwDeBH4Svdkv7aqpgGq6kVgTdd8HfB8r/uhrrYOONirH+xqx/Wp\nqteBV5JccEpHJEkaqWFnFN/rlp7WM5gdvJXBrOK4ZiMcV+ZuIkk6HVafTOOq+u8kU8B7gekka6tq\nultWeqlrdgi4sNdtfVebrd7v80KSVcC5VXX4RGPYvn3797cnJyeZnJw8mUOQpBVvamqKqampke0v\nVe2JQJIfA16rqleTvAF4BPg08G4GF6BvT/Jx4Pyq2tpdzP4s8E4GS0pfAS6rqkrydeCjwD7gy8Cf\nVtXDSW4Gfrqqbk6yGbixqjafYCw113g1Hgb3RxQQfExIbUmoqlNeqRlmRvETwM4kZzBYqvqbqnqo\ne9K/P8lHgOcY3OlEVT2V5H7gKeA14Obes/stwD3A2cBDVfVwV78buC/JAeBl4AdCQpK0OOacUSwl\nzih0lDMKaXjznVH4zmxJUpNBIUlqMigkSU0GhSSpyaCQJDUZFJKkJoNCktRkUEiSmgwKSVKTQSFJ\najIoJElNBoUkqcmgkCQ1GRSSpCaDQpLUZFBIkpoMCklSk0EhSWoyKCRJTQaFJKnJoJAkNRkUkqQm\ng0KS1GRQSJKaDApJUpNBIUlqMigkSU0GhSSpyaCQJDUZFJKkJoNCktRkUEiSmgwKSVKTQSFJajIo\nJElNBoUkqWnOoEiyPsneJN9K8mSSj3b185PsSfJ0kkeSnNfrsy3JgST7k1zTq1+V5IkkzyTZ0auf\nmWRX1+exJBeN+kAlSadmmBnFEeBjVfVW4BeBW5L8FLAVeLSqLgf2AtsAklwJbAKuAK4D7kySbl93\nAVuqagOwIcm1XX0LcLiqLgN2AHeM5OgkSfM2Z1BU1YtV9c1u+3+A/cB64AZgZ9dsJ3Bjt309sKuq\njlTVs8ABYGOSCeCcqtrXtbu316e/rweAq+dzUJKk0TmpaxRJLgbeBnwdWFtV0zAIE2BN12wd8Hyv\n26Gutg442Ksf7GrH9amq14FXklxwMmOTJC2MoYMiyZsYvNq/tZtZ1IwmM7+ej8zdRJJ0OqweplGS\n1QxC4r6qerArTydZW1XT3bLSS139EHBhr/v6rjZbvd/nhSSrgHOr6vCJxrJ9+/bvb09OTjI5OTnM\nIUjS2JiammJqampk+0vV3BOBJPcC/1lVH+vVbmdwAfr2JB8Hzq+qrd3F7M8C72SwpPQV4LKqqiRf\nBz4K7AO+DPxpVT2c5Gbgp6vq5iSbgRuravMJxlHDjFcr3+D+iAKCjwmpLQlVdcorNXMGRZJ3AV8F\nnmTwf2YBnwAeB+5nMBN4DthUVa90fbYxuJPpNQZLVXu6+juAe4CzgYeq6taufhZwH/B24GVgc3ch\nfOZYDAoBBoV0MhY8KJYSg0JHGRTS8OYbFL4zW5LUZFBIkpoMCklSk0EhSWoyKCRJTQaFJKlpxQTF\nxMTFJGFi4uLFHookrSgr5n0U3lc/Xvx9S8PzfRSSpAVlUEiSmgwKSVKTQSFJajIoJElNBoUkqcmg\nkCQ1GRSSpCaDQpLUZFBIkpoMCklSk0EhSWoyKCRJTQaFJKnJoJAkNRkUkqQmg0KS1GRQSJKaDApJ\nUpNBIUlqMigkSU0GhSSpyaCQJDUZFJKkJoNCktRkUEiSmgwKSVKTQSFJapozKJLcnWQ6yRO92vlJ\n9iR5OskjSc7rfW9bkgNJ9ie5ple/KskTSZ5JsqNXPzPJrq7PY0kuGuUBSpLmZ5gZxWeAa2fUtgKP\nVtXlwF5gG0CSK4FNwBXAdcCdSdL1uQvYUlUbgA1Jju5zC3C4qi4DdgB3zON4JEkjNmdQVNXXgP+a\nUb4B2Nlt7wRu7LavB3ZV1ZGqehY4AGxMMgGcU1X7unb39vr09/UAcPUpHIckaYGc6jWKNVU1DVBV\nLwJruvo64Pleu0NdbR1wsFc/2NWO61NVrwOvJLngFMclSRqx1SPaT41oPwBpfXP79u3f356cnGRy\ncnKEP1qSlr+pqSmmpqZGtr9TDYrpJGurarpbVnqpqx8CLuy1W9/VZqv3+7yQZBVwblUdnu0H94NC\nkvSDZr6I/uQnPzmv/Q279BSOf6W/G/hwt30T8GCvvrm7k+kS4FLg8W556tUkG7uL2x+a0eembvsD\nDC6OS5KWiFS1V42SfA6YBH4UmAZuA74IfJ7BTOA5YFNVvdK138bgTqbXgFurak9XfwdwD3A28FBV\n3drVzwLuA94OvAxs7i6En2gsNdt4B/lTQJjrmLT8+fuWhpeEqmou6zf7L6f/yQwKHeXvWxrefIPC\nd2ZLkpoMCklSk0EhSWoyKCRJTQaFJKnJoJAkNRkUkqQmg0KS1GRQSJKaDApJUpNBoWVhYuJikpCE\niYmLF3s40ljxs560LBz7/cKxDzL29y0Nw896kiQtKINCktRkUEiSmgwKSVKTQSFJajIoJElNBoUk\nqcmgkCQ1GRSSpCaDQpLUZFBoyep/vpOkxeNnPWnJ+sHPd/KznqRT4Wc9SZIWlEEhSWoyKLSkHL0u\n4d+ckJYOg2KZWql/yGd6+jmguv9KWgq8mL1MzbzQu1KOuf97HPBitjRfXsyWJC0og0KLaqUuoUkr\niUGh064fDkevScznuoRhIy0sr1EsU8v5GsWwb6Q7tn1y7ZbTuZBOB69RjJGlfuuor+yllckZxTLS\nuiNoKRzz8TOFs4HvArB27ZuBwa2va9e+ubfcBM4opIW3YmYUSd6b5F+TPJPk44s9Hh3TmsnM/r3v\n0r/24PsjpOVrSQRFkjOAPwOuBd4KfDDJTy3uqJaTsxZ0yaf/JD9zeckAWFqmpqYWewhLhudidJZE\nUAAbgQNV9VxVvQbsAm5Y5DEtI8e/eh/WbLOB1rWGUdylpIXjk+MxnovRWSpBsQ54vvf1wa42Fk7l\nCftk952EVaveOOtsYNS3rEpaOZZKUCwr/Sf21hPxsO1G/YTd/7n9fXzve/836/767SSpb0nc9ZTk\nF4DtVfXe7uutQFXV7TPaLf5gJWkZms9dT0slKFYBTwNXA98BHgc+WFX7F3VgkiRWL/YAAKrq9SS/\nA+xhsBx2tyEhSUvDkphRSJKWrmVzMXtc35CXZH2SvUm+leTJJB/t6ucn2ZPk6SSPJDlvscd6uiQ5\nI8k/JdndfT2W5yLJeUk+n2R/9/h45xifi99L8i9Jnkjy2SRnjsu5SHJ3kukkT/Rqsx57km1JDnSP\nm2uG+RnLIijG/A15R4CPVdVbgV8EbumOfSvwaFVdDuwFti3iGE+3W4Gnel+P67n4E+ChqroC+Dng\nXxnDc5HkJ4HfBa6qqp9lsKT+QcbnXHyGwXNj3wmPPcmVwCbgCuA64M4MPnunaVkEBWP8hryqerGq\nvtlt/w+wH1jP4Ph3ds12AjcuzghPryTrgfcBf9krj925SHIu8MtV9RmAqjpSVa8yhueiswp4Y5LV\nwBuAQ4zJuaiqrwH/NaM827FfD+zqHi/PAgcYPL82LZegGOs35B2V5GLgbcDXgbVVNQ2DMAHWLN7I\nTqs/Bv6A49/wMY7n4hLgP5N8pluG+4skP8wYnouqegH4I+DfGQTEq1X1KGN4LnrWzHLsM59LDzHE\nc+lyCYqxl+RNwAPArd3MYuZdCCv+roQk7wemuxlWa7q84s8Fg+WVq4A/r6qrgP9lsNwwjo+LH2Hw\nCvrNwE8ymFn8GmN4LhrmdezLJSgOARf1vl7f1cZCN51+ALivqh7sytNJ1nbfnwBeWqzxnUbvAq5P\n8m3gr4FfTXIf8OIYnouDwPNV9Q/d13/LIDjG8XHxHuDbVXW4ql4HvgD8EuN5Lo6a7dgPARf22g31\nXLpcgmIfcGmSNyc5E9gM7F7kMZ1OfwU8VVV/0qvtBj7cbd8EPDiz00pTVZ+oqouq6i0MHgN7q+o3\ngC8xfudiGng+yYaudDXwLcbwccFgyekXkpzdXZi9msHNDuN0LsLxs+zZjn03sLm7K+wS4FIGb3Bu\n73y5vI8iyXsZ3OVx9A15n17kIZ0WSd4FfBV4kmMfxvQJBr/c+xm8OngO2FRVryzWOE+3JO8Gfr+q\nrk9yAWN4LpL8HIOL+j8EfBv4TQYXdcfxXNzG4MXDa8A3gN8CzmEMzkWSzwGTwI8C08BtwBeBz3OC\nY0+yDdjC4FzdWlV75vwZyyUoJEmLY7ksPUmSFolBIUlqMigkSU0GhSSpyaCQJDUZFJKkJoNCktRk\nUEiSmv4f/8SREjZb9BYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10bf46ed0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "plt.bar(range(len(h)),h)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-49-4b620f997db4>, line 12)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-49-4b620f997db4>\"\u001b[0;36m, line \u001b[0;32m12\u001b[0m\n\u001b[0;31m    gc[i]/ float(totals[i]):\u001b[0m\n\u001b[0m                           ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "def findGCbyPos(reads):\n",
    "    gc = [0] * 100\n",
    "    totals = [0] * 100\n",
    "    \n",
    "    for read in reads:\n",
    "        for i in range(len(read)):\n",
    "            if read[i] == 'C' or read[i] == 'G':\n",
    "                gc[i] += 1\n",
    "            totals[i] += 1\n",
    "    for i in range(len(gc)):\n",
    "        if totals[i] > 0:\n",
    "            gc[i]/ float(totals[i]):\n",
    "    return gc\n",
    "gc = findGCByPos(human_reads)\n",
    "print(gc)\n",
    "\n",
    "plt.plot(range(len(gc)), gc)\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [Root]",
   "language": "python",
   "name": "Python [Root]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
